cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+        _CUDA_VRANGES::begin(__range), _CUDA_VRANGES::__unwrap_end(__range), __unwrapped_begin(), size());
+    }
+  }
+
+#ifndef _CCCL_DOXYGEN_INVOKED // doxygen conflates the overloads
+  _CCCL_TEMPLATE(class _Range)
+  _CCCL_REQUIRES(__compatible_range<_Range> _CCCL_AND _CUDA_VRANGES::forward_range<_Range> _CCCL_AND(
+    !_CUDA_VRANGES::sized_range<_Range>))
+  _CCCL_HIDE_FROM_ABI async_mdarray(const __env_t& __env, _Range&& __range)
+      : async_mdarray(
+          __env,
+          static_cast<size_type>(_CUDA_VRANGES::distance(_CUDA_VRANGES::begin(__range), _CUDA_VRANGES::end(__range))),
+          ::cuda::experimental::uninit)
+  {
+    if (size() > 0)
+    {
+      using _Iter = _CUDA_VRANGES::iterator_t<_Range>;
+      this->__copy_cross<_Iter, __detect_transfer_kind<__is_host_only, _Range>>(
+        _CUDA_VRANGES::begin(__range), _CUDA_VRANGES::__unwrap_end(__range), __unwrapped_begin(), size());
+    }
+  }
+#endif // _CCCL_DOXYGEN_INVOKED
+  //! @}
+
+  //! @addtogroup assignment
+  //! @{
+
+  //! @brief Copy-assigns a async_mdarray
+  //! @param __other The other async_mdarray.
+  //! @note Even if the old async_mdarray would have enough storage available, we may have to reallocate if the stored
+  //! memory resource is not equal to the new one. In that case no memory is allocated if \p __other is empty.
+  _CCCL_HIDE_FROM_ABI async_mdarray& operator=(const async_mdarray& __other)
+  {
+    if (this == _CUDA_VSTD::addressof(__other))
+    {
+      return *this;
+    }
+
+    const auto __count = __other.size();
+    if ((__borrow_resource() != __resource_ref_t(__other.__borrow_resource()) || (capacity() != __other.capacity())))
+    {
+      __buffer_t __new_buf{__other.get_memory_resource(), __other.get_stream(), __other.size()};
+      _CUDA_VSTD::swap(__buf_, __new_buf);
+    }
+
+    __mapping_ = __other.__mapping_;
+    this->__copy_same(__other.__unwrapped_begin(), __other.__unwrapped_end(), __unwrapped_begin());
+    __policy_ = __other.__policy_;
+    return *this;
+  }
+
+  //! @brief Move-assigns a async_mdarray
+  //! @param __other The other async_mdarray.
+  //! Clears the async_mdarray and swaps the contents with \p __other.
+  _CCCL_HIDE_FROM_ABI async_mdarray& operator=(async_mdarray&& __other) noexcept
+  {
+    if (this == _CUDA_VSTD::addressof(__other))
+    {
+      return *this;
+    }
+
+    __buf_     = _CUDA_VSTD::move(__other.__buf_);
+    __mapping_ = _CUDA_VSTD::exchange(__other.__mapping_, mapping_type());
+    __policy_  = _CUDA_VSTD::exchange(__other.__policy_, __policy_t::invalid_execution_policy);
+    return *this;
+  }
+
+  //! @brief Copy-assigns from a different async_mdarray
+  //! @param __other The other async_mdarray.
+  _CCCL_TEMPLATE(class... _OtherProperties)
+  _CCCL_REQUIRES(__properties_match<_OtherProperties...>)
+  _CCCL_HIDE_FROM_ABI async_mdarray& operator=(const async_mdarray<_Tp, _OtherProperties...>& __other)
```

Comment:
This doesn't look like it could compile, because it conflates the required template parameters `Extents` and `Layout` with the parameter pack `Properties...`.

Should we consider constraining assignment on whether the ranks match, and whether the destination has any compile-time extents that would not be compatible with the source?

Reply:
yeah that was a leftover that assumed matching types